linkerd/tls/route/src/sni.rs

```diff
@@ -0,0 +1,94 @@
+use linkerd_dns as dns;
+use linkerd_tls::ServerName;
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub enum MatchSni {
```

Comment:
It's probably worth documenting SNI matching semantics with some of the Gateway API's [verbiage](https://github.com/kubernetes-sigs/gateway-api/blob/ffe4b631bf3f0333ef6c6fc902cd4755a369bda0/apis/v1alpha2/tlsroute_types.go#L51-L57).